### PR TITLE
SoundWire: merge master add APIs

### PR DESCRIPTION
**Pull request by plbossart**
@bardliao this is a tentative alternate solution to your PR https://github.com/thesofproject/linux/pull/2027

I took the exact opposite solution and I find it a lot simpler.

### CODE REVIEW

**Comment by bardliao** [comment]
> @bardliao this is a tentative alternate solution to your PR #2027
> 
> I took the exact opposite solution and I find it a lot simpler.

@plbossart Sorry, I didn't see this PR until now, and it is quite late. So I will try it tomorrow.

**Comment by plbossart** [comment]
> > @bardliao this is a tentative alternate solution to your PR #2027
> > I took the exact opposite solution and I find it a lot simpler.
> 
> @plbossart Sorry, I didn't see this PR until now, and it is quite late. So I will try it tomorrow.

no worries. My view is that we have the add/del functions, but also the process_event and startup, so I wanted to have a consistent API for all interactions between the parent device and the soundwire master device.

**Comment by bardliao** [comment]
> no worries. My view is that we have the add/del functions, but also the process_event and startup, so I wanted to have a consistent API for all interactions between the parent device and the soundwire master device.

That is a good point. A comment from me is that where should we allocate the sdw_bus. Currently, we allocate the sdw_bus and assign the bus to master deivce (md) in link ops. However,  as you know, Vinod also request us to split the patch to a generic sdw master device patch and an adding link_ops patch. It means that the API should work without link ops to my understanding. In other word, the sdw_bus should be allocated and assigned before calling sdw_master_device_add()

File: drivers/soundwire/qcom.c [hunk 1]
```diff
@@ -877,8 +868,7 @@ static int qcom_swrm_remove(struct platform_device *pdev)
 	struct qcom_swrm_ctrl *ctrl = dev_get_drvdata(&pdev->dev);
 	int err;
 
-	sdw_delete_bus_master(&ctrl->bus);
-	err = sdw_master_device_del(ctrl->md);
+	err = sdw_master_device_del(&ctrl->md);
```
Comment (lyakh) [inline drivers/soundwire/qcom.c:871]: I think something is missing here: I don't see where the type of `ctrl->md` changed from a pointer to an embedded structure?

File: drivers/soundwire/master.c [hunk 2]
```diff
@@ -84,6 +84,18 @@ int sdw_master_device_add(struct sdw_master_device *md,
 		}
 	}
 
+	/* the bus handle may be set initially or in the link_ops, check now */
+	if (!md->bus)
+		goto link_add_err;
```
Comment (lyakh) [inline drivers/soundwire/master.c:89]: just to check, that this is really intended: if `nd->bus` is `NULL` here, we return 0 and `sdw_add_bus_master()` is never called?

**Comment by plbossart** [comment]
> > no worries. My view is that we have the add/del functions, but also the process_event and startup, so I wanted to have a consistent API for all interactions between the parent device and the soundwire master device.
> 
> That is a good point. A comment from me is that where should we allocate the sdw_bus. Currently, we allocate the sdw_bus and assign the bus to master deivce (md) in link ops. However, as you know, Vinod also request us to split the patch to a generic sdw master device patch and an adding link_ops patch. It means that the API should work without link ops to my understanding. In other word, the sdw_bus should be allocated and assigned before calling sdw_master_device_add()

the bus can be initialized either at the parent level, or in the link ops. As you said, it's required that the bus be initialized with the master_ops, port_ops and compute_params before calling sdw_bus_add_master, nothing says it has to be initialized before calling sdw_master_device_add...

File: drivers/soundwire/master.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/master.c:89]: no, nice catch. We should set ret = -EINVAL here. thanks for spotting this.

File: drivers/soundwire/qcom.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/qcom.c:871]: yes, also flagged by kbuild, I need to change this for qcom.c

**Comment by plbossart** [comment]
The tentative direction is to go with the initial solution suggested by @bardliao (use bus-centric APIs), so this PR is officially abandoned.

**Pull request closed without merging**